### ISSUE TITLE
Hi bro, when i reproduce the code on Cifar100 with 60%-Sym, the results is only 40%

### ISSUE THREAD

**Issue by LanXiaoPang613** [iTheresaApocalypse/IDO#1]
Hi bro, when i reproduce the code on Cifar100 with 60%-Sym, the results is only 40%. The attachment is the souce code. Furthermore, there are some issues in your code which will leading to error running.

[IDO_reproduce.zip](https://github.com/user-attachments/files/31935161/IDO_reproduce.zip)

**Comment by LanXiaoPang613** [comment]
This is the log of the experiments:

<img width="1371" height="632" alt="Image" src="https://github.com/user-attachments/assets/d28c0602-300e-45d4-a622-1cc914ca6a67" />

**Comment by iTheresaApocalypse** [comment]
Hi, thanks for the report and for trying to reproduce our results.
 
For CIFAR-100 with pretrained ResNet-50 + 60% Sym, ~40% is not the expected result. In Table 1 of the paper, this number is very close to the Standard CE baseline (41.12%). So this usually means Stage 2 did not run as intended, or the run fell back to a plain CE setup.
 
We re-ran this setting on our side for Stage1. Under the paper configuration (pretrained=True, input 224×224, ResNet-50, sym 0.6), Stage 1 test accuracy is already around 70% (best ~72% in our current run), which is far above 40%. The final IDO number comes from Stage 2 on top of that checkpoint. We will report that after finishing Stage 2.
 
Could you please double-check the following?
 
Both stages must be run, and Stage 2 must load the Stage 1 checkpoint + wrong-event file. Reporting only Stage 1, or continuing with CE after Stage 2 fails, will stay near the Standard baseline.
 
Please use --cfg (not --config), and the paths are config/stage_1/ and config/stage_2/ (not config/stage1/).
 
Force the same backbone for both stages. Please pass --backbone resnet50 explicitly, for example:
 ```
python [stage_1.py](https://github.com/iTheresaApocalypse/IDO/issues/stage_1.py) --cfg config/stage_1/cifar100.yaml --noise_mode sym --noise_ratio 0.6 --backbone resnet50
python [stage_2.py](https://github.com/iTheresaApocalypse/IDO/issues/stage_2.py) --cfg config/stage_2/cifar100.yaml --noise_mode sym --noise_ratio 0.6 --backbone resnet50
```
If Stage 2 is accidentally started with ViT-B, it will not load the ResNet-50 checkpoint from Stage 1.
 
Please keep pretrained: True. The dataloader uses this flag to choose the input size: pretrained → 224×224, otherwise 32×32. Training ImageNet ResNet-50 from scratch at 32×32 with 60% noise can easily drop to ~40%.
 
In the log, you should see something like:
 
Loading ResNet50, pretrained:True, stage:2
image shape: torch.Size([3, 224, 224])
successful loading of base_model/[resnet50_pretrained_cifar100_sym_0.6.pt](https://github.com/iTheresaApocalypse/IDO/issues/resnet50_pretrained_cifar100_sym_0.6.pt)
If possible, could you share the full Stage 1 / Stage 2 command lines and the corresponding logs? We are happy to help pinpoint the mismatch.

**Comment by LanXiaoPang613** [comment]
Hi bro, we use the following commands to run the IDO framework:

> BACKBONE='resnet50'

> python stage_1.py --cfg './config/stage_1/cifar100.yaml' --noise_mode 'sym' --noise_ratio 0.6 --backbone "$BACKBONE"

> python stage_2.py --cfg './config/stage_2/cifar100.yaml' --noise_mode 'sym' --noise_ratio 0.6 --backbone "$BACKBONE"

the below text is the correspongding log：
(vmamba) PS C:\Users\pc\Desktop\noisy labels论文汇总\20260105\IDO-main> bash .\run.sh
Infomation: dataset-cifar100, noise_mode-sym, noise_ratio-0.6
save noisy labels to ./data/cifar-100-python\noise_file\cifar100_sym_0.6 ...
Loading ResNet50, pretrained:True, stage:1, num_class:100
In stage 1, whether freeze the feature extractor:False
len of wrong_event:50000
image shape:torch.Size([3, 224, 224])
Epoch [  1/  5] Iter[391/391]    total-loss: 4.4980
| Train Epoch #1         Accuracy: 2.89


| Test Epoch #1  Accuracy: 8.37

Epoch [  2/  5] Iter[391/391]    total-loss: 4.3995
| Train Epoch #2         Accuracy: 5.08


last epoch prediction change: 50000 
current epoch prediction change: 43949 

| Test Epoch #2  Accuracy: 13.36

Epoch [  3/  5] Iter[391/391]    total-loss: 4.4255
| Train Epoch #3         Accuracy: 7.48


last epoch prediction change: 43949 
current epoch prediction change: 38920 

| Test Epoch #3  Accuracy: 19.31

Epoch [  4/  5] Iter[391/391]    total-loss: 4.1853
| Train Epoch #4         Accuracy: 9.75


last epoch prediction change: 38920 
current epoch prediction change: 35457 

| Test Epoch #4  Accuracy: 23.50

Epoch [  5/  5] Iter[391/391]    total-loss: 4.3291
| Train Epoch #5         Accuracy: 11.74


last epoch prediction change: 35457 
current epoch prediction change: 32919 

Epoch [  5] saves last epoch' model as the base model by calculating prediction change.
| Test Epoch #5  Accuracy: 29.20

Infomation: dataset-cifar100, noise_mode-sym, noise_ratio-0.6
Loading ResNet50, pretrained:True, stage:2, num_class:100
In stage 2, fully fine tuned the model
image shape:torch.Size([3, 224, 224])
Mixture 1, prob: 0.19, alpha1: 2.93, beta1: 1.78
Mixture 2, prob: 0.81, alpha2: 1.10, beta2: 0.00
Epoch [  1/ 10] Iter[391/391]    loss: 0.611
| Train Epoch #1         Accuracy: 10.86

Mixture 1, prob: 0.21, alpha1: 2.17, beta1: 1.26
Mixture 2, prob: 0.79, alpha2: 1.10, beta2: 0.00
None

| Test Epoch #1  Accuracy: 26.31

Epoch [  2/ 10] Iter[391/391]    loss: 0.920
| Train Epoch #2         Accuracy: 11.75

Mixture 1, prob: 0.23, alpha1: 1.77, beta1: 1.10
Mixture 2, prob: 0.77, alpha2: 1.10, beta2: 0.00
None

| Test Epoch #2  Accuracy: 28.11

Epoch [  3/ 10] Iter[391/391]    loss: 0.873
| Train Epoch #3         Accuracy: 12.80

Mixture 1, prob: 0.26, alpha1: 1.40, beta1: 1.10
Mixture 2, prob: 0.74, alpha2: 1.10, beta2: 0.00
None

| Test Epoch #3  Accuracy: 30.13

Epoch [  4/ 10] Iter[391/391]    loss: 1.110
| Train Epoch #4         Accuracy: 13.84

Mixture 1, prob: 0.27, alpha1: 1.22, beta1: 1.10
Mixture 2, prob: 0.73, alpha2: 1.17, beta2: 0.00
None

| Test Epoch #4  Accuracy: 32.95

Epoch [  5/ 10] Iter[391/391]    loss: 1.004
| Train Epoch #5         Accuracy: 14.92

Mixture 1, prob: 0.27, alpha1: 1.14, beta1: 1.10
Mixture 2, prob: 0.73, alpha2: 1.98, beta2: 0.00
None

| Test Epoch #5  Accuracy: 33.56

Epoch [  6/ 10] Iter[391/391]    loss: 1.357
| Train Epoch #6         Accuracy: 15.73

Mixture 1, prob: 0.28, alpha1: 1.08, beta1: 1.10
Mixture 2, prob: 0.72, alpha2: 3.26, beta2: 0.00
None

| Test Epoch #6  Accuracy: 35.10

Epoch [  7/ 10] Iter[391/391]    loss: 1.142
| Train Epoch #7         Accuracy: 16.49

Mixture 1, prob: 0.28, alpha1: 1.01, beta1: 1.10
Mixture 2, prob: 0.72, alpha2: 5.43, beta2: 0.00
None

| Test Epoch #7  Accuracy: 36.43

Epoch [  8/ 10] Iter[391/391]    loss: 1.021
| Train Epoch #8         Accuracy: 17.21

Mixture 1, prob: 0.28, alpha1: 0.95, beta1: 1.10
Mixture 2, prob: 0.72, alpha2: 8.81, beta2: 0.00
None

| Test Epoch #8  Accuracy: 37.05

Epoch [  9/ 10] Iter[391/391]    loss: 1.055
| Train Epoch #9         Accuracy: 17.95

Mixture 1, prob: 0.28, alpha1: 0.89, beta1: 1.10
Mixture 2, prob: 0.72, alpha2: 12.89, beta2: 0.00
None

| Test Epoch #9  Accuracy: 38.26

**Comment by LanXiaoPang613** [comment]
> Hi bro, when i reproduce the code on Cifar100 with 60%-Sym, the results is only 40%. The attachment is the souce code. Furthermore, there are some issues in your code which will leading to error running.
> 
> [IDO_reproduce.zip](https://github.com/user-attachments/files/31935161/IDO_reproduce.zip)

Moreover, the link is the source code and run.sh file we used.

**Comment by iTheresaApocalypse** [comment]
Hi, thanks for sharing the full logs and the `IDO_reproduce` package.
 
**Short answer:** Stage 1 is just standard **cross-entropy (CE)** training (collect wrong events + save a base model). Your Stage 1 already failed, so Stage 2 / BMM cannot recover. This does **not** look like an IDO algorithm failure.
 
Even under normal CE training with a pretrained ResNet-50 at 224×224, Stage 1 test accuracy should **not** be this low. Getting only ~8% after Epoch 1 (and ~29% after 5 epochs) is abnormal for CE itself, before any IDO-specific loss is involved.
 
### Why we say Stage 1 failed on your side
In your log:
- Stage 1 Epoch 1 test = **8.37%**, final ≈ **29.20%**
- Stage 1 loss stays around **4.2–4.5** (near random for 100 classes)
 
Under the same setting (`pretrained=True`, `224×224`, `frozen=False`, ResNet-50, `sym 0.6`), Stage 1 CE test accuracy should already be around **~70%**. Your later ~38–40% is close to the paper’s **Standard CE** baseline (41.12%), not IDO (81.42%).
 
### We re-ran with your uploaded code
We took your `IDO_reproduce` source + `[run.sh](https://github.com/iTheresaApocalypse/IDO/issues/run.sh)` setting, and ran it in an environment aligned with `requirements.txt` (`torch==2.0.1`, `torchvision==0.15.2`, `timm==0.9.7`). Your code itself is fine:
 
| Epoch | Your Stage1 (from your log) | Your code in our env |
|------|-----------------------------:|---------------------:|
| 1 | 8.37 | **65.68** |
| 2 | 13.36 | **72.19** |
| 3 | 19.31 | **72.73** |
| 4 | 23.50 | **72.32** |
| 5 | 29.20 | **70.29** |
 
So the gap is very likely from the **`vmamba` conda env**, not from the training script / lr / freeze / backbone settings in your zip.
 
### Our reference run (CIFAR-100, ResNet-50, sym 60%)
Environment: `torch==2.0.1`, `torchvision==0.15.2`, `timm==0.9.7` (matched to `requirements.txt`).
 
**Stage 1 (standard CE):**
| Epoch | Train | Test |
|------|------:|-----:|
| 1 | 19.62 | 66.10 |
| 2 | 30.06 | 71.17 |
| 3 | 33.06 | 70.84 |
| 4 | 35.09 | 72.17 |
| 5 | 37.32 | 69.09 |
 
**Stage 2 (IDO):**
| Epoch | Train | Test |
|------|------:|-----:|
| 1 | 33.25 | 72.18 |
| 2 | 33.92 | 73.50 |
| 3 | 35.00 | 74.26 |
| 4 | 36.19 | 76.23 |
| 5 | 37.08 | 76.55 |
| 6 | 37.77 | 77.36 |
| 7 | 38.30 | 79.02 |
| 8 | 38.77 | 79.86 |
| 9 | 39.06 | 80.11 |
| 10 | 39.15 | **80.75** |
 
**Best Acc: 80.75%** (close to paper 81.42%).
 
### BMM mixture also shows the Stage-1 problem
Initial Stage-2 BMM:
 
| | Ours | Yours |
|--|------|------|
| Mix1 (clean-ish) | 0.40 / α=0.43 / β=1.41 | 0.19 / α=2.93 / β=1.78 |
| Mix2 (noisy-ish) | 0.60 / α=1.05 / β=0.01 | 0.81 / α=1.10 / **β=0.00** |
 
Your Mixture-2 stays degenerate (`β=0`) and α later explodes (up to ~12.9). That is expected if Stage-1 wrong-event statistics are unreliable.
 
### Please check on your side
1. Create a **fresh env** from `requirements.txt` (do not reuse `vmamba`).
2. Confirm versions: `torch==2.0.1`, `torchvision==0.15.2`, `timm==0.9.7`.
3. Confirm `timm` actually downloads ImageNet ResNet-50 weights (watch HF/cache warnings).
4. After Stage 1 Epoch 1, check test accuracy: if it is still ~8–10% instead of ~60%+, Stage 1 CE is not learning and Stage 2 will stay near the CE baseline.
 
Happy to keep helping if you can share `torch` / `timm` / `torchvision` versions from the `vmamba` env and whether pretrained weights were cached successfully.

**Comment by LanXiaoPang613** [comment]
> Hi, thanks for sharing the full logs and the `IDO_reproduce` package.
> 
> **Short answer:** Stage 1 is just standard **cross-entropy (CE)** training (collect wrong events + save a base model). Your Stage 1 already failed, so Stage 2 / BMM cannot recover. This does **not** look like an IDO algorithm failure.
> 
> Even under normal CE training with a pretrained ResNet-50 at 224×224, Stage 1 test accuracy should **not** be this low. Getting only ~8% after Epoch 1 (and ~29% after 5 epochs) is abnormal for CE itself, before any IDO-specific loss is involved.
> 
> ### Why we say Stage 1 failed on your side
> In your log:
> 
> * Stage 1 Epoch 1 test = **8.37%**, final ≈ **29.20%**
> * Stage 1 loss stays around **4.2–4.5** (near random for 100 classes)
> 
> Under the same setting (`pretrained=True`, `224×224`, `frozen=False`, ResNet-50, `sym 0.6`), Stage 1 CE test accuracy should already be around **~70%**. Your later ~38–40% is close to the paper’s **Standard CE** baseline (41.12%), not IDO (81.42%).
> 
> ### We re-ran with your uploaded code
> We took your `IDO_reproduce` source + `[run.sh](https://github.com/iTheresaApocalypse/IDO/issues/run.sh)` setting, and ran it in an environment aligned with `requirements.txt` (`torch==2.0.1`, `torchvision==0.15.2`, `timm==0.9.7`). Your code itself is fine:
> 
> Epoch	Your Stage1 (from your log)	Your code in our env
> 1	8.37	**65.68**
> 2	13.36	**72.19**
> 3	19.31	**72.73**
> 4	23.50	**72.32**
> 5	29.20	**70.29**
> So the gap is very likely from the **`vmamba` conda env**, not from the training script / lr / freeze / backbone settings in your zip.
> 
> ### Our reference run (CIFAR-100, ResNet-50, sym 60%)
> Environment: `torch==2.0.1`, `torchvision==0.15.2`, `timm==0.9.7` (matched to `requirements.txt`).
> 
> **Stage 1 (standard CE):**
> 
> Epoch	Train	Test
> 1	19.62	66.10
> 2	30.06	71.17
> 3	33.06	70.84
> 4	35.09	72.17
> 5	37.32	69.09
> **Stage 2 (IDO):**
> 
> Epoch	Train	Test
> 1	33.25	72.18
> 2	33.92	73.50
> 3	35.00	74.26
> 4	36.19	76.23
> 5	37.08	76.55
> 6	37.77	77.36
> 7	38.30	79.02
> 8	38.77	79.86
> 9	39.06	80.11
> 10	39.15	**80.75**
> **Best Acc: 80.75%** (close to paper 81.42%).
> 
> ### BMM mixture also shows the Stage-1 problem
> Initial Stage-2 BMM:
> 
> Ours	Yours
> Mix1 (clean-ish)	0.40 / α=0.43 / β=1.41	0.19 / α=2.93 / β=1.78
> Mix2 (noisy-ish)	0.60 / α=1.05 / β=0.01	0.81 / α=1.10 / **β=0.00**
> Your Mixture-2 stays degenerate (`β=0`) and α later explodes (up to ~12.9). That is expected if Stage-1 wrong-event statistics are unreliable.
> 
> ### Please check on your side
> 1. Create a **fresh env** from `requirements.txt` (do not reuse `vmamba`).
> 2. Confirm versions: `torch==2.0.1`, `torchvision==0.15.2`, `timm==0.9.7`.
> 3. Confirm `timm` actually downloads ImageNet ResNet-50 weights (watch HF/cache warnings).
> 4. After Stage 1 Epoch 1, check test accuracy: if it is still ~8–10% instead of ~60%+, Stage 1 CE is not learning and Stage 2 will stay near the CE baseline.
> 
> Happy to keep helping if you can share `torch` / `timm` / `torchvision` versions from the `vmamba` env and whether pretrained weights were cached successfully.

Thank you for your reply, the issue is seems from the timm version, we update it to 0.9.7, the test acc has been normal.

**Comment by iTheresaApocalypse** [comment]
Hi, glad to hear that upgrading timm to 0.9.7 fixed the issue!
 
You're welcome — feel free to use IDO as a baseline for comparison in your work, and we'd appreciate a citation if you find it helpful. If you run into any other problems during reproduction, just let us know.
